src/Microsoft.Framework.FileSystemGlobbing/Infrastructure/MatcherContext.cs

```diff
@@ -0,0 +1,277 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Framework.FileSystemGlobbing.Abstractions;
+
+namespace Microsoft.Framework.FileSystemGlobbing.Infrastructure
+{
+    public class MatcherContext
+    {
+        public MatcherContext(Matcher matcher, DirectoryInfoBase directoryInfo)
+        {
+            Matcher = matcher;
+            DirectoryInfo = directoryInfo;
+            foreach (var pattern in matcher.IncludePatterns)
+            {
+                if (pattern.Contains == null)
+                {
+                    IncludePatternContexts.Add(new PatternContextLinearInclude(this, pattern));
+                }
+                else
+                {
+                    IncludePatternContexts.Add(new PatternContextRaggedInclude(this, pattern));
+                }
+            }
+            foreach (var pattern in matcher.ExcludePatterns)
+            {
+                if (pattern.Contains == null)
+                {
+                    ExcludePatternContexts.Add(new PatternContextLinearExclude(this, pattern));
+                }
+                else
+                {
+                    ExcludePatternContexts.Add(new PatternContextRaggedExclude(this, pattern));
+                }
+            }
+        }
+
+        public Matcher Matcher { get; }
+        public DirectoryInfoBase DirectoryInfo { get; }
+        public IList<PatternContextBase> IncludePatternContexts { get; } = new List<PatternContextBase>();
+        public IList<PatternContextBase> ExcludePatternContexts { get; } = new List<PatternContextBase>();
+        public List<string> Files { get; private set; }
```

Comment:
Clean up return types, such as using `IList<T>` on public stuff, not the concrete types.